trin-core/src/portalnet/events.rs

```diff
@@ -114,20 +112,93 @@ impl PortalnetEvents {
     // This could be handled in the UtpListener impl, but for consistency with data handling
     // and clarity it can be put here
     async fn process_utp_byte_stream(&mut self) {
-        for (_, conn) in self.utp_listener.utp_connections.iter_mut() {
+        for (_, conn) in self.utp_listener.write().await.utp_connections.iter_mut() {
             let received_stream = conn.recv_data_stream.clone();
-            if received_stream.is_empty() {
-                continue;
+
+            if UtpMessage::is_valid(&received_stream[..]) {
+                let message = UtpMessage::decode(&received_stream[..]);
+
+                match self
+                    .utp_listener
+                    .read()
+                    .await
+                    .listening
+                    .get(&conn.conn_id_recv)
+                {
+                    Some(message_type) => match message_type {
+                        // 8 is the Offer/Accept uTP stream handler
+                        8 => {
+                            let payload = UtpAccept::from_ssz_bytes(&message.payload[..])
+                                .unwrap()
+                                .message;
+
+                            for (key, content) in payload {
+                                store(key, content).unwrap();
+                            }
+                        }
+                        _ => warn!("Received uTP Stream Message of unknown type"),
+                    },
+                    _ => warn!("uTP listening HashMap doesn't have uTP stream message type"),
+                }
+
+                // message was successfully handled, remove it from the buffer
+                conn.recv_data_stream = received_stream[(message.len() as usize)..].to_owned();
             }
+        }
+    }
+}
 
-            let message_len = u32::from_be_bytes(received_stream[0..4].try_into().unwrap());
+pub struct UtpMessage {
+    length: u32,
+    payload: Vec<u8>,
+}
 
-            if message_len + 4 >= received_stream.len() as u32 {
-                // handle data function(&received_stream[4..(message_len + 4)]
+impl UtpMessage {
+    pub fn new(payload: Vec<u8>) -> Self {
+        UtpMessage {
+            length: payload.len() as u32,
+            payload: payload,
+        }
+    }
+
+    pub fn encode(&mut self) -> Vec<u8> {
+        let mut buf = vec![];
+        buf.extend_from_slice(&self.length.to_be_bytes());
+        buf.append(&mut self.payload);
+        buf
+    }
+
+    pub fn decode(bytes: &[u8]) -> UtpMessage {
```

Comment:
It seems that it would be more idiomatic to combine `decode` and `is_valid` so that `decode` returns a `Result`. Also would like to see some unit tests for `decode`.